### ISSUE TITLE
Add Plugins system support

### ISSUE THREAD

**Issue by damon-kwok** [ponylang/corral#143]
# Summary
The plug-in system is similar to a set of callbacks that allow Corral to trigger and run custom tasks  as it processes commands.

# Motivation
- Provide expansion capability for `Corral`.
- Users can contribute `Corral` without changing corral code.

# Detailed design
- The plug-in system is designed to support cross-platform
- All plug-ins must support cross-platform, which needs to be checked automatically before merging PR.

For example: An empty plug-in
| corral
├── plugins
│   ├── example
│       ├── INSTALL     # plugin installer
│       ├── COMMAND     # (optional) run user command
│       ├── on_create   # (optional)
│       ├── on_fetch    # (optional)
│       ├── on_update   # (optional)
│       ├── on_build    # (optional)
│       ├── on_run      # (optional)
│       ├── on_clean    # (optional)
│       └── on_deploy   # (optional)

**Comment by damon-kwok** [comment]
## Write a `statistic` plugin： counts code lines at build time: 

| corral
├── plugins
│   └── vcpkg
│       ├── INSTALL
│       └── on_build

### INSTALL:
``` sh
curl -L "https://statistic.sh" | bash
```

### on_build:
```sh
statistic *.pony ${project_root}
```

### output:
pony        11567
markdown    115
sh          76

**Comment by damon-kwok** [comment]
## Write a `vcpkg` plugin: handle cross-platform native deps:
| corral
├── plugins
│   └── vcpkg
│       ├── INSTALL  # install `vcpkg`, and set
│       └── COMMAND

### INSTALL:
``` sh
  github_deps  microsoft/vcpkg $CORRAL_REPO/vcpkg
  cd $CORRAL_REPO/vcpkg
  chmod +x ./bootstrap-vcpkg.sh
  if [ "$OS_TYPE" = "cygwin" ] || [ "$OS_TYPE" = "msys2" ] || [ "$OS_TYPE" = "winnt" ]; then
    ./bootstrap-vcpkg.bat
  else
    ./bootstrap-vcpkg.sh
  fi
  vcpkg_setenv
```

### COMMAND:
```sh
vcpkg $@
```

### Use `vcpkg` plugin in corral.json:
{
  "plugins": ["vcpkg"],
  "deps": [
    {
      "locator": "github.com/ponylang/valbytes.git",
      "version": "0.5.0"
    }
  ],
  "info": {}, 
  "event": [
    {
      "on_fetch": "vcpkg install openssl",
    }
  ]
}

### use openssl in pony:
```pony
use "lib:crypto"
use "lib:ssl"
```

**Comment by damon-kwok** [comment]
## Example 3: Write a `genlib` plugin
```
| corral
├── plugins
│   └── genlib
│       └── on_create
```

### on_create:
```sh
git init
git add remote genlib https://github.com/ponylang/libgen.git
git pull genlib master

for file in $(find "${TARGET_DIR}")
do
  echo "${file}"
  if ! [ -d "${file}" ]
  then
    sed -i s/"{%%REPO_OWNER%%}"/"${REPO_OWNER}"/g "${file}"
    sed -i s/"{%%REPO_NAME%%}"/"${REPO_NAME}"/g "${file}"
    sed -i s/"{%%PACKAGE%%}"/"${PACKAGE}"/g "${file}"
    sed -i s/"{%%PROJECT_DESCRIPTION%%}"/"${PROJECT_DESCRIPTION}"/g "${file}"
    sed -i s/"{%%PROJECT_STATUS%%}"/"${PROJECT_STATUS}"/g "${file}"
    sed -i s/"{%%GITHUB_USER%%}"/"${GITHUB_USER}"/g "${file}"
    sed -i s/"{%%COMMIT_EMAIL%%}"/"${COMMIT_EMAIL}"/g "${file}"
    sed -i s/"{%%COMMIT_NAME%%}"/"${COMMIT_NAME}"/g "${file}"
    sed -i s/"{%%COPYRIGHT_YEAR%%}"/"${COPYRIGHT_YEAR}"/g "${file}"
    sed -i s/"{%%COPYRIGHT_HOLDER%%}"/"${COPYRIGHT_HOLDER}"/g "${file}"
    sed -i s/"{%%COC_EMAIL%%}"/"${COC_EMAIL}"/g "${file}"
  fi
done
```

### Import `genlib` plugin to corral.json:
{
  "plugins": ["genlib"],
  "deps": []
}

### Test
```
mkdir hello
cd hello
corral init
```

results:
```
hello
├── CHANGELOG.md
├── CODE_OF_CONDUCT.md
├── CONTRIBUTING.md
├── examples
│   └── simple-example
│       └── simple-example.pony
├── gitattributes
├── gitignore
├── LICENSE
├── Makefile
├── README.md
├── RELEASE_PROCESS.md
└── STYLE_GUIDE.md
```

**Comment by SeanTAllen** [comment]
I dont see from reading this how it provides any value other than "can write things in a language other than Pony" but that opens up a mess of questions about dependencies for the other code. So to use a plugin, I might have to have bash installed. How does that work on Windows? Does the plugin need to also have a powershell version and a batch version?

At the moment, I don't see how if plugins are part of the corral repo that this is desirable given all the questions that would be raised versus writing the command in Pony.

**Comment by damon-kwok** [comment]
Yes, bash is a required dependency. We can use the idea of `OPAM`( ocaml's package manager), which will automatically download `msys2` on the windows platform. We can also use `git-bash`.

**Comment by SeanTAllen** [comment]
A plugin system that isn't in Pony also bypasses the capabilities based directory security that using only Pony provides. Commands written in Pony using the existing corral system to add commands can be limited to working on a subset of directories. By including non-pony code from ambient auth languages, that goes out the window.

**Comment by SeanTAllen** [comment]
> Yes, bash is a required dependency. We can use the idea of OPAM( ocaml's package manager), which will automatically download msys2 on the windows platform. We can also use git-bash.

This is a huge requirement and one I'm not willing to take on lightly.

**Comment by damon-kwok** [comment]
The size of msys2 compressed package is about 50mb

**Comment by SeanTAllen** [comment]
> The size of msys2 compressed package is about 50mb

The size doesn't matter. We are asking people to take on an additional currently unneeded dependency for what benefit?

**Comment by damon-kwok** [comment]
It allows users to use the common commands of unix to do anything.

**Comment by SeanTAllen** [comment]
> It allows users to use the common commands of unix to do anything.

What problem that exists now that people need solved does this address?

**Comment by damon-kwok** [comment]
I want to add a plug-in system to corral. The plug-in system can help people achieve: 
- Handle native dependencies. 
- Custom automated tasks
- ...
